Sources/FoundationEssentials/URL/URITemplate_PercentEncoding.swift

```diff
@@ -0,0 +1,244 @@
+internal import RegexBuilder
+
+extension String {
+    /// Convert to NFC and percent-escape.
+    func normalizedAddingPercentEncoding(
+        withAllowedCharacters allowed: URL.Template.Expression.Operator.AllowedCharacters
+    ) -> String {
+        switch allowed {
+        case .unreserved:
+            return addingPercentEncodingToNFC(
+                allowed: allowed
+            )
+        case .unreservedReserved:
+            return withContiguousNFCAndOutputBuffer(allowed: allowed) { input -> String in
+                switch input {
+                case .noConversionNorEncodedNeeded: return self
+                case .needsEncoding(input: let inputBuffer, outputCount: let outputCount):
+                    return withUnsafeTemporaryAllocation(of: UInt8.self, capacity: outputCount + 1) { outputBuffer -> String in
+                        addPercentEscapesForUnreservedReserved(
+                            inputBuffer: inputBuffer,
+                            outputBuffer: outputBuffer
+                        )
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// For the `unreserved / reserved / pct-encoded` case, create a String by percent encoding the NFC input as needed.
+fileprivate func addPercentEscapesForUnreservedReserved(
+    inputBuffer: UnsafeBufferPointer<UTF8.CodeUnit>,
+    outputBuffer: UnsafeMutableBufferPointer<UTF8.CodeUnit>
+) -> String {
+    let allowed = URL.Template.Expression.Operator.AllowedCharacters.unreservedReserved
+
+    var remainingInput = inputBuffer[...]
+    var outputIndex = 0
+
+    func write(_ a: UInt8) {
+        outputBuffer[outputIndex] = a
+        outputIndex += 1
+    }
+
+    while let next = remainingInput.popFirst() {
+        // Any (valid) existing escape sequences need to be copied to the output verbatim.
+        // But any `%` that are not part of a valid escape sequence, need to be encoded.
+        guard
+            next != UInt8(ascii: "%") || remainingInput.count < 2
+        else {
+            // Is this a valid escape sequence?
+            func isHex(_ a: UInt8) -> Bool {
+                (0x30 <= a && a <= 0x39) ||
+                (0x41 <= a && a <= 0x46) ||
+                (0x61 <= a && a <= 0x66)
+            }
+            if isHex(remainingInput[remainingInput.startIndex]) && isHex(remainingInput[remainingInput.startIndex + 1]) {
+                write(next)
+            } else {
+                write(UInt8(ascii: "%"))
+                write(UInt8(ascii: "2"))
+                write(UInt8(ascii: "5"))
+            }
+            continue
+        }
+        if allowed.isAllowedCodeUnit(next) {
+            write(next)
+        } else {
+            write(UInt8(ascii: "%"))
+            write(hexToAscii(next >> 4))
+            write(hexToAscii(next & 0xf))
+        }
+    }
+    write(0)
+    return String(cString: outputBuffer.baseAddress!)
```

Comment:
I think it might be preferable to do `String(decoding: outputBuffer[..<outputIndex], as: UTF8.self)` to avoid the null-terminator and cString initializer.